### ISSUE TITLE
System.Diagnostics: Activity.SetParent does not work if activity is already started

### ISSUE THREAD

**Issue by ramonsmits** [dotnet/runtime#63883]
### Description

I have a usecase where the (root) parent activity isn't known when that is started. Only until meta data retrieved is discovered the parent needs to be set. I tried doing that via `System.Diagnostics.Activity.SetParent` and invoking it didn't give any error. However, invoking this didn't have any effect.

The [System.Diagnostics.Activity.SetParent docs](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.setparentid?view=net-6.0) state:

> This method should only be used before starting the Activity object. This method has no effect if you call it after the Activity object has started.

This surprises me a bit. If that is the case I would maybe expect an `InvalidOperationException` or something because the activity is already started but primarily I don't understand why this is not possible.

It seems this is a very common scenario where a task is started like fetching some data the represents an action where that results in invoking logic.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L511-L539

### Reproduction Steps

1. Start an activity via `ActivitySource.StartActivity` that reads data, data which is created by another parent process from a tracing perspective
2. Invoke the created `Activy.SetParentId`

### Expected behavior

Ability to still set the parent ID for a started activity.

Currently it is not, either I would expect this to work or have in the documentation clearly defined why this isn't even possible if this is a limitation by the open telemetry specification.

### Actual behavior

Invoking `SetParentId` does nothing

### Configuration

net5

**Comment by tarekgh** [comment]
@ramonsmits thanks for your report. Please have a look at https://github.com/dotnet/runtime/issues/60078 for more information about the reason behind this design. I am closing this issue but feel free to send any question if there is anything clear from the other issue.

**Comment by tarekgh** [comment]
CC @noahfalk

**Comment by ramonsmits** [comment]
> @ramonsmits thanks for your report. Please have a look at https://github.com/dotnet/runtime/issues/60078 for more information about the reason behind this design. I am closing this issue but feel free to send any question if there is anything clear from the other issue.

Our need is similar to that issue. Unfortunately creating a new activity would remove the correlatio. To the current outer activity. It's this activity that should be linked to the parent as that is it's trigger
 In that context it also creates a transaction etc. all of that isn't correlated if the parent cannot be set.

Almost cannot believe this cannot be done as basically all APIs work in a way that data is received via querying, event or callback.

**Comment by tarekgh** [comment]
> Unfortunately creating a new activity would remove the correlatio. 

Can't you copy the correlation to the newly created Activity? 

> Almost cannot believe this cannot be done as basically all APIs work in a way that data is received via querying, event or callback.

The comment https://github.com/dotnet/runtime/issues/60078#issuecomment-939131319 described the reason not allowing that. 
Can you talk more about your scenario? Who creates and starts the activity in the first place? At what stage in your process is this happening? and when you receive the meta-data and want to set the parent?

**Comment by ramonsmits** [comment]
> Can't you copy the correlation to the newly created Activity?

I could, but then the trace misses full details.


> The comment [#60078 (comment)](https://github.com/dotnet/runtime/issues/60078#issuecomment-939131319) described the reason why not allowing that.

Yes I read that "The runtime makes a promise to those listeners that the TraceID and SpanID observable in those callbacks will be fixed for the lifetime of the Activity."

I understand that, there is also no reason to change these. I only want to change the `ParentID`. Maybe the  TraceID and SpanID share some "trace" identificiation. I could understand that but that seems like an oversight.



> Can you talk more about your scenario? Who creates and starts the activity in the first place? At what stage in your process is this happening? and when you receive the meta-data and want to set the parent?

Sure, lets take a message broker type of scenario:

- Process A, sends message X based on a UI action. Meaning, HTTP POST results in sending a messages to a broker. The post created a trace which represents the starting point of the trace.
- Process B, receives message X, does some transactional IO in the context of that data and completes.

The problem is with process B where "receiving" should be part of the trace. The act of receiving is already a trace that is started, then when the data is received it contains meta data containing the trace parent ID.

Lets say that the receive is transaction, meaning BEGIN TRAN and COMMIT TRAN are IO operations that are meaningful to trace.

We have a "receive message" activity that will do a "BEGIN TRAN", receive data, read trace parent ID but cannot assign it apparently. Assume it create a new activity called "process message" and completes. Now we want to COMMIT and want to trace its duration in this context. This is not possible and can only be a child under "receive message". 

Meaning, now exist 2 traces "trees" if a new "process message" activity would be created and cannot be correlated while they IMHO should.

- One that starts with "HTTP POST", showing the sending and processing of data
- One that starts with "receive message" with underneath *only* the commit and not the sub-tree "process message" as that has become part of the parent tree

Hence the need to assign the parent of the "receive message" activity at a later moment.

One thing I have not tried but maybe it possible is a "LazyActivity".

Meaning, 
1. call `ActivitySource.CreateActivity` instead of `ActivitySource.StartActivity`
2. call `.SetStartTime(...)` (in the hope this will not start the activity
3. receive the data, parse the parent trace ID, invoke `.SetParentId(...)` which hopefully is possible because its not started yet|

Only problem that might exist here is that calling calling `CreateActivity` will not assign `Activity.Current` which is problematic as then the activities that represent the actual IO operations begin the transaction and receive the message data will not have a 'parent'. It seems that if I need to go that route that I need to use `CreateActivity` and then somehow assign the 'parent' activity object but that does not exist. There is only a `.SetParentId` and not a `.SetParent(Activity)`.Maybe I could assign `Activity.Current myself..` but I guess that I cannot call `ActivitySource.StartActivity` as that might depend on the fact that any `Activity` object in `Activity.Current` is already "started".

**Comment by ramonsmits** [comment]
@tarekgh Code says more then a thousand words

```c#
var ActivitySource = new ActivitySource("Test");
string traceId;

using (var produce = ActivitySource.StartActivity("Produce", ActivityKind.Producer))
{
    traceId = produce.Id;
}

// For diagnostics tracing the whole task is desirable to include the fetch operation
using (var outer = ActivitySource.StartActivity("Consume", ActivityKind.Consumer))
{
    // Start transaction here

    // Should be part of the trace as the operation could take a while due to unknown technical reasons
    using (var fetch = ActivitySource.StartActivity("Fetch"))
    {
        // Read data and obtain traceID
    }

    // Simulate retrieval of traceId
    outer.SetParentId(traceId).Start();
    using (var inner = ActivitySource.StartActivity("Process")) // We could assign traceId here, but then the activities part of the parent aren't visible. Showing the parent activity separately.
    {
        // Consume
    }

    using (var commit = ActivitySource.StartActivity("Commit"))
    {
    }
}
```

**Comment by ramonsmits** [comment]
@tarekgh do you understand the use-case?

**Comment by noahfalk** [comment]
Most scenarios that I have seen ignore the computational time prior to receiving incoming data so the Activity only includes the work after the incoming message was received. However if you wanted to capture that earlier work you could connect the Activities together using a link:

```C#
var ActivitySource = new ActivitySource("Test");

// For diagnostics tracing the whole task is desirable to include the fetch operation
using (var outer = ActivitySource.StartActivity("Consume"))
{
    // Should be part of the trace as the operation could take a while due to unknown technical reasons
    string incomingTraceId = null;
    using (var fetch = ActivitySource.StartActivity("Fetch"))
    {
        // Read data and obtain incomingTraceId 
    }

    using (var inner = ActivitySource.StartActivity("Process", ActivityKind.Consumer, incomingTraceId, 
        links: new ActivityLink[] { new ActivityLink(Activity.Current.Context) } )) 
    {
        // Consume

        using (var commit = ActivitySource.StartActivity("Commit"))
        {
        }
    }
}
```

> I only want to change the ParentID

The [OpenTelemetry specification](https://opentelemetry.io/docs/reference/specification/trace/api/#span-operations) wouldn't allow that to happen. The parent span context has to be specified when a Span is started. Although I didn't notice any text that explicitly says the Span context is immutable after creation I believe that is the intent and the common understanding among all implementers. It is also implied by the lack of any defined API that would mutate span context. From a practical standpoint samplers treat root spans and child spans differently and anything that allowed a root span to transform into child span after it was started would break samplers.

OpenTelemetry offers links as a more flexible mechanism to associate two spans outside of the rigid parent-child relationship and that seems like it could be a good way to capture this precursor work. Hope that helps!

**Comment by ramonsmits** [comment]
I tried the below workaround by using `CreateActivity` instead of `StartActivity` to postpone starting but do capture the start time of the activity. It runs without any internal validation errors. However, both Jeager and Application Insights do not visualize the trace and Produce, Consume and Fetch are disconnected.

```c#
var ActivitySource = new ActivitySource("Test");
string parentId;

using (var produce = ActivitySource.StartActivity("Produce", ActivityKind.Producer))
{
    parentId = produce?.Id;
}

// For diagnostics tracing the whole task is desirable to include the fetch operation
using (var outer = ActivitySource.CreateActivity("Consume", ActivityKind.Consumer))
{
    outer?.SetStartTime(DateTime.UtcNow); // Defer start until we know the parentId, but do capture start time
    // Start transaction

    // Should be part of the trace as the operation could take a while due to unknown technical reasons
    using (var fetch = ActivitySource.CreateActivity("Fetch", ActivityKind.Internal))
    {
        fetch?.SetStartTime(DateTime.UtcNow);

        // Read data and obtain traceID

        outer?.SetParentId(parentId);
        outer?.Start(); // This will set Activity.Current

        fetch?.SetParentId(outer.Id);
        fetch?.Start(); // This will set Activity.Current
    }// This will stop fetch, and replace previous activity

    // Simulate retrieval of traceId
    using (var inner = ActivitySource.StartActivity("Inner"))
    {
    }

    using (var commit = ActivitySource.StartActivity("Commit"))
    {
    }
}
```

**Comment by ramonsmits** [comment]
@noahfalk thanks for that info, I'll look into it.

**Comment by noahfalk** [comment]
> However, both Jeager and Application Insights do not visualize the trace and Produce, Consume and Fetch are disconnected

I'm not sure why offhand. I made a slight modification of your sample to log all the IDs to the console:
```
ActivityListener listener = new ActivityListener()
{
    ShouldListenTo = _ => true,
    Sample = (ref ActivityCreationOptions<ActivityContext> aco) => ActivitySamplingResult.AllDataAndRecorded,
    ActivityStopped = a => Console.WriteLine($"Name={a.DisplayName} Id={a.Id} ParentSpanId={a.ParentSpanId}")
};
System.Diagnostics.ActivitySource.AddActivityListener(listener);
``` 

Which shows the logged IDs as:
```
Name=Produce Id=00-7a21261e6b448844f7e242b79bab1fc9-46cd4e83cad24d6e-01 ParentSpanId=0000000000000000
Name=Fetch Id=00-7a21261e6b448844f7e242b79bab1fc9-3578e8cfcc444059-01 ParentSpanId=01be6dc5bab97e37
Name=Inner Id=00-7a21261e6b448844f7e242b79bab1fc9-95b113ba8aa1ea1e-01 ParentSpanId=01be6dc5bab97e37
Name=Commit Id=00-7a21261e6b448844f7e242b79bab1fc9-cf714cf288b147a1-01 ParentSpanId=01be6dc5bab97e37
Name=Consume Id=00-7a21261e6b448844f7e242b79bab1fc9-01be6dc5bab97e37-01 ParentSpanId=46cd4e83cad24d6e
```

Which gives the tree below when I match up the IDs manually:
```
Produce -> Consume -> Fetch
                   -> Inner
                   -> Commit
```

I could guess those tools are somehow unhappy with the timestamps or the Producer/Consumer span kinds but that is pure speculation. As best I can tell the data captured in your example is a valid tree.

**Comment by ramonsmits** [comment]
@noahfalk I tested with `ActivityLink` but that doesn't really solve it. It keeps the 2 traces as is but 'links' it. In Jeager I get actually a link symbol, in Application Insights I get a sort of combined visualization. Jaeger is pretty useless

Jaeger:

![image](https://user-images.githubusercontent.com/152998/150133822-9ba03949-f195-4b5a-aa0a-dd09857428d8.png)

Application Insights:

![image](https://user-images.githubusercontent.com/152998/150134101-138a7d6e-8c0c-4667-b11c-eec94abf9ba8.png)

**Comment by ramonsmits** [comment]
@noahfalk 

> I could guess those tools are somehow unhappy with the timestamps or the Producer/Consumer span kinds but that is pure speculation. As best I can tell the data captured in your example is a valid tree.

Yeah, that seems to be the case as Jaeger gives the following warning although I don't understand it: 

> invalid parent span IDs=1e7740f421bcfccc; skipping clock skew adjustment

Would you assume that the following snippets should give the same results?

Using `StartActivity`:
```c#
        using (var produce = ActivitySource.StartActivity("Produce", ActivityKind.Producer))
        {
            traceId = produce.Id;
        }

        using (var outer = ActivitySource.StartActivity("Consume", ActivityKind.Consumer, traceId))
        {
        }
```

Deferred start using `CreateActivity`:
```c#
        using (var produce = ActivitySource.StartActivity("Produce", ActivityKind.Producer))
        {
            traceId = produce.Id;
        }

        using (var outer = ActivitySource.CreateActivity("Consume", ActivityKind.Consumer))
        {
            outer.SetParentId(traceId);
            outer.Start();
        }
```

**Comment by noahfalk** [comment]
> Would you assume that the following snippets should give the same results?

Yep

> Yeah, that seems to be the case as Jaeger gives the following warning although I don't understand it...

You could try your workaround without including the SetStartTime() to see if that changes anything. Unforetunately one of the challenges of doing things differently than most others have done them is that you are more likely to encounter bugs, implementation limitations, or less intuitive visualizations. As best I can tell both your workaround and the one I suggested are capturing expected+legal telemetry but the backends still aren't drawing a nice picture.

**Comment by ramonsmits** [comment]
StartActivity immediately invokes CreateActivity:

 https://github.com/dotnet/runtime/blob/e54df94bc9a5e1e064835be817d3e9ac773e1d9a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L125-L126

Except behavior is different when `parentId` is null/not null.

I would assume `Activity.Start()` and `ActivitySource.CreateActivity` with a non-null `parentId` would more or less do the same things but it seems that is not the case.

- Activity.Start : https://github.com/dotnet/runtime/blob/e54df94bc9a5e1e064835be817d3e9ac773e1d9a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L623
- ActivitySource.CreateActivity : https://github.com/dotnet/runtime/blob/e54df94bc9a5e1e064835be817d3e9ac773e1d9a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L167


What I find interesting is that `Activity.Start()` is not invoked anywhere in `ActivitySource` which is what I would expect.

**Comment by ramonsmits** [comment]
> You could try your workaround without including the SetStartTime() to see if that changes anything.

Yeah, that would be similar to my `CreateActivity` and results the same `invalid parent span IDs=1e7740f421bcfccc; skipping clock skew adjustment` message. 

To me it seems the data transmitted is incorrect. Not sure what I can at the moment and also don't know if I found a bug or a valid limitation.

@noahfalk @tarekgh Any tips on what could I do from here?

**Comment by tarekgh** [comment]
> I would assume Activity.Start() and ActivitySource.CreateActivity with a non-null parentId would more or less do the same things but it seems that is not the case.

It does the same at the end. CreateActivity will just call `Activity.Create(`
https://github.com/dotnet/runtime/blob/e54df94bc9a5e1e064835be817d3e9ac773e1d9a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L285

And `Activity.Create will call `Activity.Start`

https://github.com/dotnet/runtime/blob/bbbfba31b61a2299c434172f709716325c073ef8/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L1094

Note the `CreateActivity` which calling `Activity.Create` takes a boolean flag telling if need to start the activity. this flag would be true only if the calls coming from `StartActivity` only.

Is it possible to try to defer the start of the fetch and commit activities too? I mean the code

```C#
using (var inner = ActivitySource.StartActivity("Inner"))
    {
    }

    using (var commit = ActivitySource.StartActivity("Commit"))
    {
    }
```

can use `CreateActivity` instead of `StartActivity` and then start the outer Activity first then start inner followed by commit? Maybe this will make Jager happy?

**Comment by noahfalk** [comment]
> To me it seems the data transmitted is incorrect

Is there a particular code snippet and measurement technique you are using to draw this conclusion? From what I observed earlier with your workaround that was delaying the start of Activities + pre-dating the start time, the Activity properties to got set as you intended them to be. I'm not sure if you were looking at a different example now or what output you were looking at to decide the results are incorrect. If you are looking at the final output in Jaeger or AppInsights to make that determination that includes a several intermediate layers. We will probably want to measure at different stages to narrow in on what layer is causing trouble:

1. Using ActivityListener to log Activity properties directly to the console is a good way to observe the .NET code on its own.
2. If (1) looked correct then we can use OpenTelemetry with the console exporter to test if the SDK is doing anything unexpected.
3. If (1) and (2) look correct then the issue presumably lies in a backend specific OpenTelemetry exporter or the backend itself.

**Comment by ramonsmits** [comment]
> Note the `CreateActivity` which calling `Activity.Create` takes a boolean flag telling if need to start the activity. this flag would be true only if the calls coming from `StartActivity`.

@tarekgh As mentioned earlier CreateActivity behaves different from when a parentId is set:

https://github.com/dotnet/runtime/blob/e54df94bc9a5e1e064835be817d3e9ac773e1d9a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs#L182

That if branch seems to contain code when instead later calling `SetParentId()` followed by `Start()`.

> Is it possible to try to defer the start of the fetch and commit activities too? I mean the code

As mentioned earlier I tried using `CreateActivity` as shown in the snippet: https://github.com/dotnet/runtime/issues/63883#issuecomment-1016447139

Viewing the both gives a different result in Jaeger and Application Insights where using the `CreateActivity` `SetParentID` `Start` workaround results in the message  `invalid parent span IDs=1e7740f421bcfccc; skipping clock skew adjustment message.` when looking in the details.

**Comment by ramonsmits** [comment]
> Is there a particular code snippet and measurement technique you are using to draw this conclusion?

@noahfalk Yes, because using the `CreateActivity` `SetParentID` `Start` workaround results in the message  `invalid parent span IDs=1e7740f421bcfccc; skipping clock skew adjustment message.` when looking in the details.

When only using `StartActivity` the traces visualize as expected.

> From what I observed earlier with your workaround that was delaying the start of Activities + pre-dating the start time, the Activity properties got set as you intended them to be. I'm not sure if you were looking at a different example now or what output you were looking at to decide the results are incorrect.

See https://github.com/dotnet/runtime/issues/63883#issuecomment-1016447139

> 1. Using ActivityListener to log Activity properties directly to the console is a good way to observe the .NET code on its own.

Not entirely sure what is meant here.


Here a screen capture when only using `StartActivity`:

![image](https://user-images.githubusercontent.com/152998/150808970-9130c2ae-ea62-44b7-a303-03bc7d5112e7.png)



Here a Jaeger screen capture when using `CreateActivity`, `SetParentId`, `Start`:

![image](https://user-images.githubusercontent.com/152998/150809212-e57a4291-4953-4e37-8b3d-3cac6456dca0.png)

and

![image](https://user-images.githubusercontent.com/152998/150809270-0f4b8c45-05cb-4a24-a137-c90649b80ca8.png)

**Comment by tarekgh** [comment]
>  As mentioned earlier CreateActivity behaves different from when a parentId is set:
> That if branch seems to contain code when instead later calling SetParentId() followed by Start().

I am not sure why you are concerned with the code more than the result. If you call CreateActivity and passing ParentId, you'll get the Activity created with that Id. Could you clarify what exactly the problem you are seeing with this? 


```C#
            const string w3cId = "00-99d43cb30a4cdb4fbeee3a19c29201b0-e82825765f051b47-01";
            using ActivitySource aSource = new ActivitySource("MySource");
            using ActivityListener listener = new ActivityListener();
            listener.ShouldListenTo = (activitySource) => object.ReferenceEquals(aSource, activitySource);
            listener.SampleUsingParentId = (ref ActivityCreationOptions<string> activityOptions) => ActivitySamplingResult.AllDataAndRecorded;
            listener.Sample = (ref ActivityCreationOptions<ActivityContext> activityOptions) => ActivitySamplingResult.AllDataAndRecorded;
            ActivitySource.AddActivityListener(listener);

            Activity a = aSource.CreateActivity("myActivity", ActivityKind.Client, w3cId);

            Console.WriteLine($"{a.DisplayName} ... {a.ParentId}");
            a.Start();
            Console.WriteLine($"Started: {a.DisplayName} ... {a.ParentId} ... {a.Duration}");
```

This is simply print the following which is correct. 

```
myActivity ... 00-99d43cb30a4cdb4fbeee3a19c29201b0-e82825765f051b47-01
myActivity ... 00-99d43cb30a4cdb4fbeee3a19c29201b0-e82825765f051b47-01
```

**Comment by noahfalk** [comment]
> > Using ActivityListener to log Activity properties directly to the console is a good way to observe the .NET code on its own.

> Not entirely sure what is meant here.

I was refering to the code sample I [had used earlier](https://github.com/dotnet/runtime/issues/63883#issuecomment-1016443119). Tarek's example above is similar. You can add this code to the entrypoint of any app and it will log every Activity the app creates at the point the Activity is stopped.
```
ActivityListener listener = new ActivityListener()
{
    ShouldListenTo = _ => true,
    Sample = (ref ActivityCreationOptions<ActivityContext> aco) => ActivitySamplingResult.AllDataAndRecorded,
    ActivityStopped = a => Console.WriteLine($"Name={a.DisplayName} Id={a.Id} ParentSpanId={a.ParentSpanId}")
};
System.Diagnostics.ActivitySource.AddActivityListener(listener);
```
Can you add this to your code, and show what it prints together with the Jaeger error message?
The advantage of this is that it lets us see the behavior of the .NET code independent of anything the OpenTelemetry SDK or a particular exporter is doing in its processing. It also provides a clue to decoding the error message you are getting from the Jaeger exporter because when Jaeger says ParentId=xyzxyzxyz is invalid we currently don't know which Activity that refers to. If you log all Activities to the console we should be able to see exactly which one Jaeger is unhappy about and maybe guess why Jaeger objects to it.

I also spotted https://github.com/open-telemetry/opentelemetry-dotnet/issues/2796. I don't know if it is related beyond having the same error message but highlighting it just in case.

**Comment by ramonsmits** [comment]
> Can you add this to your code, and show what it prints together with the Jaeger error message?

@noahfalk I'll do that early next week.
